### PR TITLE
[AIRFLOW-7033] Change dag and task state meanwhile

### PR DESCRIPTION
**Pull request by zhongjiajie**
In dag page, have make task instance state to success
or failed bottom, when we click those bottom not only
change task instance state but also dag run state
if possible.

* When trigger make_failed bottom, the dag run state
  should change to failed
* When trigger make_success bottom, then dag run
  state should change to success if all task instance
  state are success, otherwise should keep the
  original state

---
Issue link: [AIRFLOW-7033](https://issues.apache.org/jira/browse/AIRFLOW-7033)

Make sure to mark the boxes below before creating PR: [x]

- [x] Description above provides context of the change
- [x] Commit message/PR title starts with `[AIRFLOW-NNNN]`. AIRFLOW-NNNN = JIRA ID<sup>*</sup>
- [x] Unit tests coverage for changes (not needed for documentation changes)
- [x] Commits follow "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)"
- [x] Relevant documentation is updated including usage instructions.
- [x] I will engage committers as explained in [Contribution Workflow Example](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#contribution-workflow-example).

<sup>*</sup> For document-only changes commit message can start with `[AIRFLOW-XXXX]`.

---
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).
Read the [Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines) for more information.

### CODE REVIEW

**Comment by zhongjiajie** [comment]
Also and make_success/make_failed to dag.html/tree.html

File: airflow/api/common/experimental/mark_tasks.py
Comment (zhongjiajie) [inline airflow/api/common/experimental/mark_tasks.py:317]: Could we remove this code?

File: airflow/api/common/experimental/mark_tasks.py
Comment (zhongjiajie) [inline airflow/api/common/experimental/mark_tasks.py]: And also this one?

**Review by zhongjiajie — COMMENTED** [review]
I'm not really sure could we remove this code, it seem unnecessary

**Comment by zhongjiajie** [comment]
> I'm not really sure could we remove this code, it seem unnecessary

Oh, now I realize the code is not unnecessary

File: airflow/api/common/experimental/mark_tasks.py
Comment (zhongjiajie) [reply airflow/api/common/experimental/mark_tasks.py]: I change the code in the next commit

**Comment by zhongjiajie** [comment]
@kaxil @potiuk @mik-laj @nuclearpinguin PTAL. Just ignore the conflict, will fix it after review.

**Comment by zhongjiajie** [comment]
Fix conflict

**Comment by zhongjiajie** [comment]
PTAL @ashb @kaxil @potiuk @mik-laj @turbaszek @feluelle

File: airflow/api/common/experimental/mark_tasks.py
Comment (potiuk) [reply airflow/api/common/experimental/mark_tasks.py:317]: Not sure. @mik-laj -> maybe you can take a look, you know much better now how changing state works and whether this change will be ok.

**Comment by potiuk** [comment]
It looks good to me, I just would like to have second opinion (@mik-laj @kaxil @ashb )

**Review by potiuk — APPROVED** [review]

**Comment by zhongjiajie** [comment]
I almost forget about this PR 😢

**Comment by feluelle** [comment]
This does not look right to me - to put this button on task-level actions window. Wouldn’t it be better to put it on the DagRun actions window (when you click on the round colored button)?

**Comment by zhongjiajie** [comment]
> This does not look right to me - to put this button on task-level actions window. Wouldn’t it be better to put it on the DagRun actions window (when you click on the round colored button)?

Hi @feluelle , we already have set state to `failed/running/success` in dagrun page. And this feature not trigger from dagrun, but trigger from Dags page when trying to make task failed/success, so I think it should not put in the DagRun page.

This patch will check if all task in dagrun success or not, if all task success will change dagrun state to success, otherwise will keep in to failed.

But I agree with you that it maybe confuse users, with dag in task windows, maybe we should make `include_dag` default `True` in `_mark_task_instance_state`

```py
        return self._mark_task_instance_state(dag_id, task_id, origin, execution_date,
                                              confirmed, upstream, downstream,
                                              future, past, State.FAILED)
                                              future, past, include_dag=True, State.FAILED)
```
and change the `UPDATING.md`. I'm left here wait for other committer advices.

**Comment by zhongjiajie** [comment]
ping @kaxil @mik-laj @feluelle again, I need one more approval/review here

**Comment by feluelle** [comment]
No, I mean on the tree view page. So you can click on one dag run and can mark this dagrun and maybe future or past ones (as succeeded or failed).

**Comment by zhongjiajie** [comment]
@feluelle You'r right, I that another way to mark dag run state. But this PR is try to fix the exists bug(maybe) in tool view in dag/tree view for task instance. The core idea is "should also change dag run state if all task instance success or at least one fail". So I think this two function do not conflict

**Comment by zhongjiajie** [comment]
BTW, I just checkout and find out we already have this function in tree view dagrun
Screen Shot 2020-04-15 at 00 09 15

**Comment by feluelle** [comment]
Yes that’s the window I meant.

Don‘t you think it would make more sense to add an optional button „task instances“ next to mark success/fail button? instead of „dag“ on the tsk instance dialog window?

**Comment by kaxil** [comment]
> Yes that’s the window I meant.
> 
> Don‘t you think it would make more sense to add an optional button "task instances" next to mark success/fail button? instead of "dag" on the task instance dialog window?

I agree with @feluelle , I think it would make more sense to not have this on Task Instance Modal dialog and instead on a specific DAG.

**Comment by zhongjiajie** [comment]
Ok, I will try to
* Remove "DAG" button in Task Instance dialog in Graph view
* Add task instances in dag run dialog in Tree view

But I insist on add function `_set_dag_run_state_according_ti` in `mark_tasks.py` and use `include_dag=True` as default value in `view.py`, that mean mark dagrun success if all task instance success as default behavior
```diff
        return self._mark_task_instance_state(dag_id, task_id, origin, execution_date,
                                              confirmed, upstream, downstream,
                                              future, past, State.FAILED)
-                                             future, past, State.FAILED)
+                                             future, past, include_dag=True, State.FAILED)
```
Cause I think It's a missing function in Task Instance Modal dialog, I have some situation about this dialog. For example
```py
task1 >> [task2, task3]
```
* **CASE ONE**: task2 and task3 failed, and task3 **not need** to re-run(not related about time and next scheduler task work fine), but task2 **have to be rerun**. I just mark task3 success and clear task2 include downstream. every thing work fine and dagrun mark as success after finish task2.
* **CASE TWO**: task2 and task3 failed and both of them **not need to rerun**, I just click task1 and mark task instance success include downstream, but dagrun still with failed status. Yeah we could go to tree view mark dagrun success after I add optional button beside 'mark success/failed' button, but it's mean I have to move from graph to tree view, which meaningless step.

So I think change dagrun state in Task Instance Modal dialog if possible make sense, and I want to keep it this function. And remove the "DAG" button in Task Instance Modal dialog to avoid misleading to user. WDYT @feluelle @kaxil @potiuk

**Comment by zhongjiajie** [comment]
And BTW, I alway check DAG view at first, because in big dag with many tasks, I think it's more straightforward about the relationship in dag view instead of tree view.

**Comment by zhongjiajie** [comment]
No stale, sorry about that, I got some health problem recently, will continue after recover from illness.

**Pull request closed without merging**
